src/main/protobuf/remote_output_service.proto

```diff
@@ -0,0 +1,305 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package remote_output_service;
+
+import "build/bazel/remote/execution/v2/remote_execution.proto";
+import "google/protobuf/empty.proto";
+
+option java_package = "com.google.devtools.build.lib.remote";
+option java_outer_classname = "RemoteOutputServiceProto";
+option go_package = "remoteoutputservice";
+
+// The Remote Output Service may be used by users of the Remote
+// Execution API to construct a directory on the local system that
+// contains all output files of a build.
+//
+// Primitive implementations of this API may simply download files from
+// the Content Addressable Storage (CAS) and store them at their
+// designated location. Complex implementations may use a pseudo file
+// system (e.g., FUSE) to support deduplication, lazy loading and
+// snapshotting.
+service RemoteOutputService {
+  // Methods that can be invoked at any point in time.
+
+  // Clean all data associated with a single output path, so that the
+  // next invocation of StartBuild() yields an empty output path. This
+  // may be implemented in a way that's faster than removing all of the
+  // files from the file system manually.
+  rpc Clean(CleanRequest) returns (google.protobuf.Empty);
+
+  // Signal that a new build is about to start.
+  //
+  // The client uses this call to obtain a directory where outputs of
+  // the build may be stored, called the output path. Based on the
+  // parameters provided, the remote output service may provide an empty
+  // output path, or one that has contents from a previous build of the
+  // same project.
+  //
+  // In case the output path contains data from a previous build, the
+  // remote output service is responsible for calling
+  // ContentAddressableStorage.FindMissingBlobs() for all of the objects
+  // that are stored remotely. This ensures that these objects don't
+  // disappear from the Content Addressable Storage while the build is
+  // running. Any files that are absent must be removed from the output
+  // path and reported through InitialOutputPathContents.modified_paths.
+  rpc StartBuild(StartBuildRequest) returns (StartBuildResponse);
+
+  // Methods that can only be invoked during a build.
+
+  // Create one or more files, directories or symbolic links in the
+  // output path.
+  rpc BatchCreate(BatchCreateRequest) returns (google.protobuf.Empty);
+
+  // Obtain the status of one or more files, directories or symbolic
+  // links that are stored in the input path.
+  rpc BatchStat(BatchStatRequest) returns (BatchStatResponse);
+
+  // Signal that a build has been completed.
+  rpc FinalizeBuild(FinalizeBuildRequest) returns (google.protobuf.Empty);
+}
+
+message CleanRequest {
+  // The output base identifier that was provided to
+  // StartBuildRequest.output_base_id whose data needs to be removed.
+  string output_base_id = 1;
+}
+
+message StartBuildRequest {
```

Comment:
Here is a redacted version of the internal message, for reference. If this seems compatible, we could try to open-source the relevant parts, although it may be a struggle to make copybara do what we need.

```
message StartBuildRequest {
  // The FUSE daemon workspace id for the workspace in which the build is performed.
  required string workspace = 1;
  // Root of the directory under which Bazel builds its entire output tree.
  required string blaze_exec_root = 2;

  // The uid and gid of the user performing the build.
  // The 'uid' must match the user making this request.
  required int32 uid = 3;
  required int32 gid = 4;

  // Root directory of the workspace in which Bazel is running
  optional string workspace_root = 7;

  // Build invocation ID in Bazel terminology
  optional string invocation_id = 9;
}

message StartBuildReply {
  // Build id will always be set if StartBuildRequest succeeds.
  required string build_id = 1;

  // Relative path of artifacts in this workspace that have been externally
  // modified since the last StartBuildReply for the same workspace (e.g.
  // "bazel-out/gcc-.../bin/some/file.o").
  optional StatusProto modified_path_status = 3;
  repeated string modified_path = 4;
}
```

Reply:
Nice! Thanks for sharing! A couple of questions I have:

- `StartBuildRequest.{uid,gid}`: This is likely sent across because the daemon that Google has developed internally is multi-user, right? In my case I'm building a single-user daemon, creating a cache directory and FUSE mount in the user's home directory. Both are valid approaches, I guess. I've had some good experiences using `SO_PEERCRED` (Linux ) / `LOCAL_PEERCRED` (BSD) in combination with grpc-go. This allows a gRPC server to obtain the PID, UID and GID of the calling process in an authenticated way. It may therefore be possible to implement a multi-user daemon without encoding this information explicitly, right?
- `StartBuildReply.modified_path`: (How) does this mechanism deal with FUSE daemon restarts/crashes/loss of data? For the protocol in this PR I made it so that the FUSE daemon also sends back the invocation of the previous build, which should allow Bazel to detect state inconsistencies.

I'm always a fan of open sourcing code, but I suppose it may be a good idea to let Bazel use its own protocol that is strongly based on REv2. Being able to support REv2 instance names and use the Digest{,Function}, Output{File,Directory,Symlink} messages seems nice. What are your thoughts?

Reply:
Re the modified path question, the "modified_path_status" field says whether the modified_path list can be trusted. If it says no, we assume ModifiedFileSet.EVERYTHING_MODIFIED and fall back to manual stats.

Agreed that it may make sense to have this protocol reflect Bazel's feature set. The code that could be shared is fairly minimal, but I think it'll be nice to keep things as parallel as possible, both for future readers and for any convergence efforts. I must admit that I'm not sure how layering these protos would work, so actual sharing may be more trouble than it's worth.

Reply:
That part makes complete sense, but what I want to aim at specifically is that the protocol that I came up with also sends back the UUID of the previous build. In my opinion this should be a requirement to make this robust, right?

Google's development environment is pretty much streamlined: folks just run Blaze on their gLinux workstations where all of this stuff is readily deployed. In the case of Bazel, people are most likely going to do things that are a bit crazy: run the FUSE daemon in one container, and run Bazel in another one. Those containers may be permanent, they may be short lived, etc.

The negotiation process between the FUSE daemon and Bazel on whether to reuse existing output path state is incomplete if the inputs are:

- blaze_exec_root
- invocation_id

And the outputs are merely:

- modified_path_status
- modified_path

Because what will happen if someone accidentally runs two containerized copies of Bazel against a single FUSE daemon that happen to use the same (virtualized!) blaze_exec_root? The FUSE daemon could respond 'modified_path_status == OK', even though the previous build could have been triggered by another instance of Bazel. By letting the FUSE daemon return the previous invocation ID to Bazel, Bazel can detect these cases and ignore the FUSE daemon's state if needed.

Reply:
Furthermore, the protocol that you pasted above lets the FUSE daemon return a build_id, which I suppose is some UUID/identifier generated by the FUSE daemon that Bazel must attach to Google's internal equivalent of BatchStat() / BatchCreate(), right?

The protocol in this PR differs from that in that it doesn't let the FUSE daemon generate any identifier. The invocation ID provided by Bazel is reused for that. Do we think that's acceptable?

Reply:
I'm not super familiar with the build id returned, and I'll probably bow out and let @alexjski investigate the Google side soon, but from what I can see, the "build id" is actually a directory on the filesystem that the FUSE daemon controls, and so it's only used for setting up the symlink properly. I think that corresponds to "output_path_prefix" in this code.

I agree that in the scenario where multiple Bazels are writing to the same execroot, things can get crazy. I'm not sure, though, that anything is guaranteed in that scenario, since I *think* that means the two Bazel instances are sharing the same output base, which seems like it won't end well for anybody. Having the previous id in the protocol seems like a nice defensive measure, though, agreed.

Reply:
I know that extracting the UID using `{LOCAL,SO}_PEERCRED` works with grpc-go, both on Linux and macOS. I am using exactly this for some other project and it's working great.

If it's necessary to pass in authentication/identification information explicitly, I don't think there's really a need to extend the protocol to accommodate this. Couldn't users just use `--remote_header` for this, similar to how people can pass in authentication tokens?

Reply:
Considering that I have first-hand experience using `LOCAL_PEERCRED` on macOS and `SO_PEERCRED` on Linux, and that `--remote_header` also exists, let me resolve this thread.

Reply:
About the `build_id` in the response in the internal protocol -- that seems a little bit like abuse of the same term -- this has nothing to do with a Bazel build id, it is something that the output service manages and corresponds to a workspace. I agree with Janak that it is conceptually similar to the `output_path_prefix`.